src/Mvc/Mvc.Analyzers/src/TagHelpersInCodeBlocksAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+        }
+
+        private DiagnosticDescriptor TagHelperInCodeBlockDiagnostic { get; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                var symbolCache = new SymbolCache(context.Compilation);
+
+                if (symbolCache.TagHelperRunnerRunAsyncMethodSymbol == null)
+                {
+                    // No-op if we can't find bits we care about.
+                    return;
+                }
+
+                InitializeWorker(context, symbolCache);
+            });
+        }
+
+        private void InitializeWorker(CompilationStartAnalysisContext context, SymbolCache symbolCache)
+        {
+            context.RegisterOperationBlockStartAction(startBlockContext =>
+            {
+                startBlockContext.RegisterOperationAction(context =>
+                {
+                    var awaitOperation = (IAwaitOperation)context.Operation;
+
+                    if (awaitOperation.Operation.Kind != OperationKind.Invocation)
+                    {
+                        return;
+                    }
+
+                    var invocationOperation = (IInvocationOperation)awaitOperation.Operation;
+
+                    if (!IsTagHelperRunnerRunAsync(invocationOperation.TargetMethod, symbolCache))
+                    {
+                        return;
+                    }
+
+                    var parent = context.Operation.Parent;
+                    while (parent != null && !IsParentMethod(parent))
+                    {
+                        parent = parent.Parent;
+                    }
+
+                    if (parent == null)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = (IMethodSymbol)(parent switch
+                    {
+                        ILocalFunctionOperation localFunctionOperation => localFunctionOperation.Symbol,
+                        IAnonymousFunctionOperation anonymousFunctionOperation => anonymousFunctionOperation.Symbol,
+                        IMethodBodyOperation methodBodyOperation => startBlockContext.OwningSymbol,
+                        _ => null,
+                    });
+
+                    if (methodSymbol == null)
+                    {
+                        // Unsupported operation type.
+                        return;
+                    }
+
+                    if (!methodSymbol.IsAsync ||
+                        !symbolCache.TaskType.IsAssignableFrom(methodSymbol.ReturnType))
+                    {
+                        if (parent.Syntax.GetDiagnostics().Any(diagnostic => diagnostic.Id == TagHelperInCodeBlockDiagnostic.Id))
```

Comment:
Did you actually see this happen? I would be really surprised if this works. I wouldn't expect the syntax tree to update DURING your analyzer running.

Reply:
Didn't fully test this part. You be right though. This totally fails to work as expected 😄 

![image](https://user-images.githubusercontent.com/2008729/57954010-9585f800-78a6-11e9-8bc1-c775bd61eb61.png)